### PR TITLE
✨  enables storage policy in failure domain 

### PR DESCRIPTION
**Pull request by RenilRaj-BR**
**What this PR does:**
Currently failure domain doesn't support storage policy and it supports specifying a datastore only. We want to add a feature  to enable specifying a storage policy for the failure domain. This provides flexibility and we don't have to specify the exact datastore name. It can also target multiple datastores when we specify a storage policy instead of a specific datastore

### CODE REVIEW

File: apis/v1alpha3/vspherefailuredomain_types.go [hunk 1]
```diff
@@ -78,6 +78,11 @@ type Topology struct {
 	// virtual machine is created/located.
 	// +optional
 	Datastore string `json:"datastore,omitempty"`
+
+	// StoragePolicy is the name of the policy that is used to target a datastore
+	// in which the virtual machine is created/located
+	// +optional
+	StoragePolicy string `json:"storagePolicy,omitempty"`
```
Comment (chrischdi) [inline apis/v1alpha3/vspherefailuredomain_types.go:85]: We should only add this to v1beta1.

File: controllers/vspherecluster_reconciler.go [hunk 2]
```diff
@@ -359,7 +359,12 @@ func (r *clusterReconciler) reconcileDeploymentZones(ctx context.Context, cluste
 }
 
 func (r *clusterReconciler) reconcileClusterModules(ctx context.Context, clusterCtx *capvcontext.ClusterContext) (reconcile.Result, error) {
-	if feature.Gates.Enabled(feature.NodeAntiAffinity) {
+	// We want to enable cluster module only when NodeAntiAffinity flag is true and failure domains are not used.
+	// When failure domain is used and an empty value is passed for data center or the resource pool in a vm template,
+	// cluster module creation will fail as there are no values in dc or resource pool.
+	// These values in the vm template gets replaced at a later stage by the values in the failure domain.
+	// When failure domain is used, we do not need the cluster module
+	if feature.Gates.Enabled(feature.NodeAntiAffinity) && clusterCtx.VSphereCluster.Spec.FailureDomainSelector == nil {
```
Comment (chrischdi) [inline controllers/vspherecluster_reconciler.go:367]: You note:

> [...] During the creation of cluster module, it validates the values in the vm template like data-center, resource-pool etc. When a failure domain is used, these values in the vm template can be empty and it get replaced by the values in the failure domain/deployment zone at a later stage. So, when the value of data-center, resource-pool etc are empty in vm template, the cluster module creation fails and cluster module doesn't make sense when failure domains are being used. To mitigate this issue, we have introduced a check in the vSphere cluster reconciler to enable cluster module creation only when the node anti-affinity is enabled and the failure domains are not being used

Is it **always** the case that cluster modules don't make sense or are there cases where they make sense?

Comment (RenilRaj-BR) [reply controllers/vspherecluster_reconciler.go:367]: @chrischdi,
when failure domain is used it is not making sense as the cluster module creation happens much before the vm template override function. The override function will replace the values in vm template with values in the failure domain. So control plane vms can end up in different zones(vSphere cluster - not vSphereCluster CR) with different values for resource pool, vm folder etc whereas cluster module gets created from the template with wrong values for resource pool, folder etc.

When failure domains are not used, cluster module makes sense

Comment (chrischdi) [reply controllers/vspherecluster_reconciler.go:367]: But it still makes sense if the different zones are in the same vcenter right?

Comment (chrischdi) [reply controllers/vspherecluster_reconciler.go:367]: @RenilRaj-BR it feels like this change is orthogonal to adding support for Storage Policies per Failure Domain.

Please consider to move this to a separate PR/issue so we can discuss on this topic properly.

**Comment by chrischdi** [comment]
CC @lubronzhan , @neolit123 , as this falls into govmomi territory :-)

File: controllers/vspherecluster_reconciler.go (hunk 2 above)
Comment (RenilRaj-BR) [reply controllers/vspherecluster_reconciler.go:367]: sure. I will move it out

File: apis/v1alpha3/vspherefailuredomain_types.go (hunk 1 above)
Comment (RenilRaj-BR) [reply apis/v1alpha3/vspherefailuredomain_types.go:85]: @chrischdi ,
I tried adding the storage policy only in v1beta1, but the conversion-gen fails to generate the `zz_generated.conversion.go` without compilation error due to missing field.
Could you please suggest

File: internal/webhooks/vspherefailuredomain.go [hunk 3]
```diff
@@ -61,6 +61,11 @@ func (webhook *VSphereFailureDomainWebhook) ValidateCreate(_ context.Context, ra
 		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "Topology", "ComputeCluster"), "cannot be empty if Hosts is not empty"))
 	}
 
+	// We should either pass a datastore or a storage policy, not both at the same time
+	if obj.Spec.Topology.Datastore != "" && obj.Spec.Topology.StoragePolicy != "" {
```
[the remark this answers is not in the document]
Comment (lubronzhan) [reply internal/webhooks/vspherefailuredomain.go:65]: Yeah looks like the priority of these two are handled at the vm creation level https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/a2bac7b1984fac89adc4f8ae54a5aeea5b828c11/pkg/services/govmomi/vcenter/clone.go#L206-L274

[the remark this answers is not in the document]
Comment (RenilRaj-BR) [reply internal/webhooks/vspherefailuredomain.go]: removed the validation as it is handled in the vm creation phase

**Comment by rikatz** [comment]
/lgtm

From a provisioning perspective, looks good :) 

Will leave for CAPV approvers to do the final review.

Thanks!

**Comment by lubronzhan** [comment]
LGTM

File: apis/v1beta1/vspherefailuredomain_types.go [hunk 4]
```diff
@@ -85,6 +85,11 @@ type Topology struct {
 	// virtual machine is created/located.
 	// +optional
 	Datastore string `json:"datastore,omitempty"`
+
+	// StoragePolicy is the name of the policy that is used to target a datastore
+	// in which the virtual machine is created/located
```
Comment (neolit123) [inline apis/v1beta1/vspherefailuredomain_types.go:90]: missing dot at the end.

File: config/default/crd/bases/infrastructure.cluster.x-k8s.io_vspherefailuredomains.yaml [hunk 5]
```diff
@@ -368,6 +368,11 @@ spec:
                     items:
                       type: string
                     type: array
+                  storagePolicy:
+                    description: |-
+                      StoragePolicy is the name of the policy that is used to target a datastore
+                      in which the virtual machine is created/located
```
Comment (neolit123) [inline config/default/crd/bases/infrastructure.cluster.x-k8s.io_vspherefailuredomains.yaml:374]: same

Comment (RenilRaj-BR) [reply config/default/crd/bases/infrastructure.cluster.x-k8s.io_vspherefailuredomains.yaml]: fixed

File: apis/v1beta1/vspherefailuredomain_types.go (hunk 4 above)
Comment (RenilRaj-BR) [reply apis/v1beta1/vspherefailuredomain_types.go]: fixed

**Review by neolit123 — COMMENTED** [review]
this is a small change, so please squash the commits to 1.

**Review by neolit123 — COMMENTED** [review]
/lgtm

**Comment by rikatz** [comment]
/lgtm
Thanks

File: apis/v1alpha4/zz_generated.conversion.go [hunk 6]
```diff
@@ -896,14 +896,10 @@ func autoConvert_v1beta1_Topology_To_v1alpha4_Topology(in *v1beta1.Topology, out
 	out.Hosts = (*FailureDomainHosts)(unsafe.Pointer(in.Hosts))
 	out.Networks = *(*[]string)(unsafe.Pointer(&in.Networks))
 	out.Datastore = in.Datastore
+	// WARNING: in.StoragePolicy requires manual conversion: does not exist in peer-type
```
Comment (chrischdi) [inline apis/v1alpha4/zz_generated.conversion.go:899]: Don't we need to do the manual conversion for this? (in both api versions)

In both vspherefailuredomain_conversion.go functions we have to do things similar to:

For the ConvertTo function:

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/apis/v1alpha4/vspheremachine_conversion.go#L33-L48

For the ConvertFrom function:

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/apis/v1alpha4/vspheremachine_conversion.go#L61-L64

**Comment by fabriziopandini** [comment]
/hold

> Currently failure domain doesn't support storage policy and it supports specifying a datastore only.

It seems that it is not required to have different storagePolicy on each failure domain, and the storagePolicy setting in the vsphere machine template already cover this use case

**Comment by rikatz** [comment]
I have discussed internally with @RenilRaj-BR and we agreed this can be closed for now.

/close

**Pull request closed without merging**